### PR TITLE
fix: workflow.status is now set properly in metrics. Fixes #8895

### LINKED ISSUE #8895: `{{workflow.status}}` is not working in metrics
## Checklist

 

* [x] Double-checked my configuration.
* [x] Tested using the latest version.
* [ ] Used the Emissary executor.

## Summary

What happened/what you expected to happen?

Running [this](https://github.com/argoproj/argo-workflows/blob/master/examples/dag-custom-metrics.yaml) example workflow for custom metrics at the workflow & template levels on the latest version (v3.3.6) publishes the workflow templates without status label!

Workflow metrics (`playground_workflow_duration`, `playground_workflow_result_counter`) are visible in prometheus without the status label => `{{workflow.status}}` is not giving any value.
While template metrics (`playground_workflow_duration_task_seconds`, `playground_workflow_result_task_counter`) with status label of `{{status}}` is working fine.

What version are you running?

3.3.6

This previously worked on version 3.2.9

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.

a shorter version of [this](https://github.com/argoproj/argo-workflows/blob/master/examples/dag-custom-metrics.yaml) example

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: dag-task-
spec:
  entrypoint: dag-task
  metrics: # Custom metric workflow level
    prometheus:
      - name: playground_workflow_new
        help: "Count of workflow execution by result status  - workflow level"
        labels:
          - key: "playground_id_workflow_counter"
            value: "test"
          - key: status
            value: "{{workflow.status}}"
        counter:
          value: "1"
  templates:
  - name: dag-task
    dag:
      tasks:
      - name: TEST-ONE
        template: echo

  - name: echo
    container:
      image: alpine:3.7
      command: [echo, "hello"]
```

```bash
# Logs from the workflow controller:
kubectl logs -n argo deploy/workflow-controller | grep ${workflow} 

# If the workflow's pods have not been created, you can skip the rest of the diagnostics.

# The workflow's pods that are problematic:
kubectl get pod -o yaml -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

# Logs from in your workflow's wait container, something like:
kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by dpadhiar**
Signed-off-by: Dillen Padhiar <dillen_padhiar@intuit.com>

Fixes #8895 

In Argo workflows v3.3.6, it was found that running an example workflow for custom metrics at the workflow & template levels would publish `{{workflow.status}}` as an empty value instead of showing the correct status. This is because the global runtime parameters are not accurately reflected by the stored workflow spec.

Example workflow (original issue):

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: dag-task-
spec:
  entrypoint: dag-task
  metrics: # Custom metric workflow level
    prometheus:
      - name: playground_workflow_duration
        help: "Duration gauge by workflow level"
        labels:
          - key: "playground_id_workflow"
            value: "test"
          - key: status
            value: "{{workflow.status}}"
        gauge:
          realtime: false
          value: "{{workflow.duration}}"
      - name: playground_workflow_result_counter
        help: "Count of workflow execution by result status  - workflow level"
        labels:
          - key: "playground_id_workflow_counter"
            value: "test"
          - key: status
            value: "{{workflow.status}}"
        counter:
          value: "1"
  templates:
  - name: dag-task
    dag:
      tasks:
      - name: TEST-ONE
        template: echo
        arguments:
          parameters: 
            - name: message
              value: "console output-->TEST-{{item.command}}"
            - name: tag
              value: "{{item.tag}}"
        withItems:
          - { tag: TEST-ONE-A, command: ONE-A }
          - { tag: TEST-ONE-B, command: ONE-B }

      - name: TEST-TWO
        template: echo
        arguments:
          parameters: 
            - name: message
              value: "console output-->TEST-{{item.command}}"
            - name: tag
              value: "{{item.tag}}"
        withItems:
          - { tag: TEST-TWO-A, command: TWO-A }
          - { tag: TEST-TWO-B, command: TWO-B }

  - name: echo
    inputs:
      parameters:
      - name: message
      - name: tag
    metrics: # Custom metric template level
      prometheus:
        - name: playground_workflow_duration_task_seconds
          help: "Duration gauge by task name in seconds - task level"
          labels:
            - key: "playground_task_name"
              value: "{{inputs.parameters.tag}}"
            - key: status
              value: "{{status}}"
          gauge:
            realtime: false
            value: "{{duration}}"
        - name: playground_workflow_result_task_counter
          help: "Count of task execution by result status  - task level"
          labels:
            - key: "playground_task_name_counter"
              value: "{{inputs.parameters.tag}}"
            - key: status
              value: "{{status}}"
          counter:
            value: "1"
    container:
      image: alpine:3.7
      command: [echo, "{{inputs.parameters.message}}"]
```

Completed workflow spec:
```yaml
spec:
  activeDeadlineSeconds: 300
  arguments: {}
  entrypoint: dag-task
  metrics:
    prometheus:
    - gauge:
        realtime: false
        value: '{{workflow.duration}}'
      help: Duration gauge by workflow level
      labels:
      - key: playground_id_workflow
        value: test
      - key: status
        value: Succeeded
      name: playground_workflow_duration
    - counter:
        value: "1"
      help: Count of workflow execution by result status  - workflow level
      labels:
      - key: playground_id_workflow_counter
        value: test
      - key: status
        value: Succeeded
      name: playground_workflow_result_counter
```

### CODE REVIEW

**Review by tczhao — COMMENTED** [review]
perhaps we should add something to the `workflow/controller/operator_metrics_test.go` as well

**Comment by sarabala1979** [comment]
@dpadhiar can you add test?

**Comment by dpadhiar** [comment]
> @dpadhiar can you add test?

Yes, will do. @tczhao Will add in metrics test.

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -3509,9 +3509,17 @@ func (woc *wfOperationCtx) setExecWorkflow(ctx context.Context) error {
...
+
+	// runtime value will be set after the substitution, otherwise will not be reflected from stored wf spec
+	woc.setGlobalRuntimeParameters()
+
 	return nil
 }
 
+func (woc *wfOperationCtx) setGlobalRuntimeParameters() {
+	woc.globalParams[common.GlobalVarWorkflowStatus] = string(woc.wf.Status.Phase)
+}
```
Comment (tczhao) [inline workflow/controller/operator.go:3521]: I think the duration should be part of this function as well
https://github.com/argoproj/argo-workflows/pull/8939/files#diff-f321d4af83fcf8311dc80c0d50c59ac4c240f761206e7bb652709870eb9feb43L247-L252

File: workflow/controller/operator_metrics_test.go [hunk 2]
```diff
@@ -783,3 +783,137 @@ func TestControllerRestartWithRunningWorkflow(t *testing.T) {
...
+  metrics:
+    prometheus:
+    - counter:
+        value: "1"
+      help: Count of workflow execution by result status  - workflow level
+      labels:
+      - key: playground_id_workflow_counter
+        value: test
+      - key: status
+        value: Succeeded
```
Comment (tczhao) [inline workflow/controller/operator_metrics_test.go:804]: I think here should be "{{workflow.status}}"?

File: workflow/controller/operator_metrics_test.go [hunk 3]
```diff
@@ -783,3 +783,137 @@ func TestControllerRestartWithRunningWorkflow(t *testing.T) {
...
+      type: Pod
+  phase: Succeeded
+  progress: 1/1
+  resourcesDuration:
+    cpu: 10
+    memory: 6
+  startedAt: "2022-06-10T20:29:42Z"
+`
+
+func TestRuntimeMetrics(t *testing.T) {
```
Comment (tczhao) [inline workflow/controller/operator_metrics_test.go:901]: I don't think this is testing the orignal bug.

You can write wf.yaml without the status
do operate twice with a pod phase change in between
and check metric before and after the 2nd operate to make sure it is changing
```
	makePodsPhase(ctx, woc, apiv1.PodSucceeded)
	woc = newWorkflowOperationCtx(woc.wf, controller)
	woc.operate(ctx)
```

**Review by tczhao — CHANGES_REQUESTED** [review]

File: workflow/controller/operator.go (hunk 1 above)
Comment (dpadhiar) [reply workflow/controller/operator.go:3521]: Will add.

**Comment by dpadhiar** [comment]
@tczhao When I am testing with `{{workflow.status}}` in the workflow, this is what error I am seeing with this test:

```
Error Trace:	operator_metrics_test.go:920
Error:      	Expected value not to be nil.
Test:       	TestRuntimeMetrics
```

The line it is failing on is `metric := controller.metrics.GetCustomMetric(metricDesc)` as it is returning nil metrics
However printing metricDesc shows it is as this string: `playground_workflow_new{playground_id_workflow_counter=test,status={{workflow.status}},}`

**Comment by tczhao** [comment]
Curious, I ran the failed e2e test locally and it worked fine

**Review by tczhao — APPROVED** [review]

**Review by sarabala1979 — APPROVED** [review]

**Comment by rvignesh89** [comment]
Is there any plan to release this in the upcoming release? I noticed this issue since v3.3.1 and have been waiting for the fix to be merged. Thanks!

**Pull request merged**